linera-faucet/server/src/lib.rs

```diff
@@ -182,72 +214,282 @@ where
+            context,
+            faucet_storage,
+            pending_requests,
+            request_notifier,
+        }
+    }
+
+    /// Runs the batch processor loop
+    async fn run(&mut self, cancellation_token: CancellationToken) {
+        loop {
+            tokio::select! {
+                _ = self.request_notifier.notified() => {
+                    if let Err(e) = self.process_batch().await {
+                        tracing::error!("Batch processing error: {}", e);
+                    }
+                }
+                _ = cancellation_token.cancelled() => {
+                    // Process any remaining requests before shutting down
+                    if let Err(e) = self.process_batch().await {
+                        tracing::error!("Final batch processing error: {}", e);
+                    }
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Processes batches until there are no more pending requests in the queue.
+    async fn process_batch(&mut self) -> Result<(), anyhow::Error> {
+        loop {
+            let max_batch_size = self.config.max_batch_size;
+            let mut batch_requests = Vec::new();
+
+            // Collect requests from the queue.
+            {
+                let mut requests = self.pending_requests.lock().await;
+                while batch_requests.len() < max_batch_size && !requests.is_empty() {
+                    if let Some(request) = requests.pop_front() {
+                        batch_requests.push(request);
+                    }
+                }
+            }
+
+            if batch_requests.is_empty() {
+                return Ok(());
+            }
+
+            tracing::info!("Processing batch of {} requests", batch_requests.len());
+
+            if let Err(err) = self.execute_batch(batch_requests).await {
+                tracing::error!("Failed to execute batch: {}", err);
+                return Err(err);
+            }
+        }
+    }
+
+    /// Executes a batch of chain creation requests
+    async fn execute_batch(
+        &mut self,
+        batch_requests: Vec<PendingRequest>,
+    ) -> Result<(), anyhow::Error> {
+        // Pre-validate: check rate limiting and existing chains
+        let mut valid_requests = Vec::new();
+
+        for request in batch_requests {
+            // Check if this owner already has a chain
+            {
+                let storage = self.faucet_storage.lock().await;
+                if let Some(existing_description) = storage.get_chain(&request.owner) {
+                    let _ = request.responder.send(Ok(existing_description.clone()));
+                    continue;
+                }
+            }
+
+            valid_requests.push(request);
+        }
 
-        if self.start_timestamp < self.end_timestamp {
+        if valid_requests.is_empty() {
+            return Ok(());
+        }
+
+        // Rate limiting check for the batch
+        let client = self
```

Comment:
Do we need to recreate the client every time we want to propose a block? It's always for the same chain.

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/4336/commits/166e8cb6ee11a2e28d8bf395f5a910c2241a84b8.